pkg/client/client.go

```diff
@@ -132,132 +151,202 @@ func (c *Client) List(ctx context.Context, namespace string, result runtime.Obje
+// UpdateStatus will then attempt to unmarshal the updated object from the response into the provide result object.
+// If the returned response object is of type Status and has .Status != StatusSuccess, the
+// additional information in Status will be used to enrich the error.
+func (c *Client) UpdateStatus(ctx context.Context, namespace string, obj, result runtime.Object, opts UpdateOptions) error {
 	defer c.setKind(result)
-	ctx, cancel := c.setupCtx(ctx, 0)
+	ctx, cancel := c.setupCtx(ctx)
 	defer cancel()
 	m, err := meta.Accessor(obj)
 	if err != nil {
 		return err
 	}
-	err = c.RESTClient.Put().
+	req := c.RESTClient.Put().
 		Prefix(c.prefix...).
 		NamespaceIfScoped(namespace, c.Namespaced).
 		Resource(c.resource).
 		Name(m.GetName()).
 		SubResource("status").
-		VersionedParams(&opts, metav1.ParameterCodec).
-		Body(obj).
-		Do(ctx).
-		Into(result)
-	return
+		VersionedParams(&opts.UpdateOptions, metav1.ParameterCodec).
+		Body(obj)
+
+	for _, modifier := range opts.RequestModifiers {
+		if modifier != nil {
+			modifier(req)
+		}
+	}
+
+	return req.Do(ctx).Into(result)
 }
 
-func (c *Client) Delete(ctx context.Context, namespace, name string, opts metav1.DeleteOptions) error {
-	ctx, cancel := c.setupCtx(ctx, 0)
+// Delete will attempt to delete the resource with the matching name in the given namespace (if client.Namespaced is set to true).
+func (c *Client) Delete(ctx context.Context, namespace, name string, opts DeleteOptions) error {
+	ctx, cancel := c.setupCtx(ctx)
 	defer cancel()
-	return c.RESTClient.Delete().
+	req := c.RESTClient.Delete().
 		Prefix(c.prefix...).
 		NamespaceIfScoped(namespace, c.Namespaced).
 		Resource(c.resource).
 		Name(name).
-		Body(&opts).
-		Do(ctx).
-		Error()
+		Body(&opts.DeleteOptions)
+
+	for _, modifier := range opts.RequestModifiers {
+		if modifier != nil {
+			modifier(req)
+		}
+	}
+
+	return req.Do(ctx).Error()
 }
 
-func (c *Client) DeleteCollection(ctx context.Context, namespace string, opts metav1.DeleteOptions, listOpts metav1.ListOptions) error {
-	ctx, cancel := c.setupCtx(ctx, 0)
+// DeleteCollection will attempt to delete all resource the given namespace (if client.Namespaced is set to true).
+func (c *Client) DeleteCollection(ctx context.Context, namespace string, opts DeleteOptions, listOpts ListOptions) error {
+	ctx, cancel := c.setupCtx(ctx)
 	defer cancel()
 	var timeout time.Duration
 	if listOpts.TimeoutSeconds != nil {
 		timeout = time.Duration(*listOpts.TimeoutSeconds) * time.Second
 	}
-	return c.RESTClient.Delete().
+	req := c.RESTClient.Delete().
 		Prefix(c.prefix...).
 		NamespaceIfScoped(namespace, c.Namespaced).
 		Resource(c.resource).
-		VersionedParams(&listOpts, metav1.ParameterCodec).
+		VersionedParams(&listOpts.ListOptions, metav1.ParameterCodec).
 		Timeout(timeout).
-		Body(&opts).
-		Do(ctx).
-		Error()
+		Body(&opts.DeleteOptions)
+
+	for _, modifier := range opts.RequestModifiers {
```

Comment:
I would recommend that we apply the delete options second. I would expect that we would want those to take precedence in a conflict between the deleteOpts and the listOpts.

Reply:
Yes I added a function comment as well to address the caller about this.